### ISSUE TITLE
Javascript visualisation in spark-notebook 

### ISSUE THREAD

**Issue by Gerrrr** [andypetrella/spark-notebook#322]
## Idea
Huge effort was put by community into the development of Javascript visualisation libraries. Having possibility to use them in spark-notebook would dramatically increase the possibilities to represent the data. Although there are bindings for D3, WISP, Rickshaw, the general way to execute Javascript would make it possible to use any libraries the user want without any additional efforts made for their bindings support.

## Development steps

1. Function which imports external javascript libraries from repl, e.g. `jsImport("url/to/lib.js")`
2. Function which
	* accepts js code and scala objects as arguments
	* executes the code on front-end
	* outputs js error if any
3. Common way to show graphics on the front-end, e.g. function, which returns id of the `output` div for the particular input block. Having this div user can add generated graphics into it.

---
Related to https://github.com/FRosner/spawncamping-dds/issues/111

**Comment by andypetrella** [comment]
okay, sounds good, then you might take a look o the `:javascript` context.
It allows you to use put any javascript you like including `require([], ...)` that would allow you to include any js.

The string after `:javascript` is interpolated hence, you could use scala types in-there so you can use stuffs like `${scalaList.mkString(",")}` in it.

What's missing is to easily use the `JsonCodec` + `ToPoint` typeclass to generate Json from types

Other missing stuff is giving a reference to the output area's element...

You may also be interested in looking at `JsWorld` that is also a good candidate where we could, within a cell, create an instance of it with extra args like:
* javascripts from the outside world (this is **missing**)
* `snippets` should work with javascript strings that defines a function
* `content` will be placed in the resulting output area

Since the data you pass in will create an observable the javascript functions you'll pass can make use of this to be also reactive.

Just some food for thinking, we can discuss what is a simple approach.

Something I'd like to keep is to keep the code to produce in as much as we can in Scala (so `:javascript` isn't really abstracting much there ^^)

**Comment by andypetrella** [comment]
@Gerrrr I createc a full example: https://github.com/andypetrella/spark-notebook/blob/master/notebooks/misc/Using%20Javascript%20and%20Scala%20to%20add%20Reactive%20Chart%20type%20LIVE%20-%20D3%20StreamGraph.snb

it's not easy but demonstrate that almost everything is there but needs easier api maybe?

also note that the function could use require to load extra libs of course

**Comment by Gerrrr** [comment]
@andypetrella Thank you for the example!
Based on it, I am going to refine the proposal and hack the small prototype during the weekend.

**Comment by Gerrrr** [comment]
## Current workflow to create custom js chart
* define js as a string. In the example it is a function, which accepts data, container and other arguments as parameters
* import
	* `notebook.front.widgets._`
	* `notebook.front.widgets.magic._`
	* `notebook.front.widgets.magic.Implicits._`
* create case class, which extends package `notebook.front.Chart` (from `package notebook.front.package.scala`). This class parametrizes container and defines the logic of passing parameters to js. Js function itself is passed as a string.
* snippets, scripts (????)
* create instance of case class, which outputs as a chart

## What is good already?
* Scala and JS code are separated.
* Explicit declaration of logic behind passing scala data structure to js
* Explicit js container configuration
* Automatic  passing container object to js function
* Possibility to use external js libraries via `requirejs`
* Reactive changes **(!)**

## What can be better?
* Use of js, defined in the browser tab, instead of passing it as a string
* Simplified API

## Improved general workflow
* define js function with `:javascript` context with 3 arguments - `data`, `container`, `options`. Alternatively import module using `requirejs`
* import 1 namespace, e.g. `notebook.front._` **(?)**
* call 1 function, e.g. `renderChart(data, function, [options], [sizes], [serializer])`, where
	* `data` – JSON string or serialazable data structure
	* `function` – function name from current js namespace
	* `options` (optional) – serialaizable data structure, which is passed as a 3rd argument to the function
	* `sizes` (optional) – tuple of ints, which specifies container size
	* `serializer` (optional) – function, which serialize `data` into JSON. Used when custom types are used.
	
## How will it work with DDS?
1. with `requirejs` import  `dds` js visualisation library, which will be at some point modularized out from the main repo
2. import main dds package
3. call `dds.bar` scala function on RDD, which will produce Servable case class
4. serialize servable into JSON
5. call `renderChart(serializedServable, "dds.draw")`

	
## Steps to be done in Spark-notebook
Implement `renderChart` function which
* serializes `data`
* find js function to call
* creates instance of the case class, which extends the `Chart` abtract class


## Questions
1. Can you please describe in more detail you idea about using JsonCodes and ToPoint for serialization from types?
2. If you would like, we can discuss the proposal in the hangouts to refine the details

**Comment by FRosner** [comment]
It would be awesome to have a mechanism to include any JS library (and connect it to scala code) without having to worry about dependency conflicts with spark-notebook itself. Something like a JS sandbox in this regard. This way we don't even have to rely on the libraries to be AMD compatible.

**Comment by andypetrella** [comment]
@Gerrrr great write up (and awesome code dig ^^).

The proposal is getting closer, like it so far. Let's see how to add ease of usage and functionalities without removing others.

# Prelude
So, there is a question for `script` and `snippet`.

`script` refers to a js file available on the server that can be loaded directly based on its name.

`snippet` will be used the same way but it's actually a full js `String`, that can either define a function or a function and an option 
```javascript
{
  f: function(d, c, o) {..}, 
  o: {
         ...
       }
}`

# Regarding the questions:
1. `JsonCodec` and `ToPoint`:
* `JsonCodec` is a simplification of Codec, where on side is always Json. Helpful for communication on the socket between server and browser in the widgets.
* `ToPoint` is a typeclass that defines how a List can always be converted to a structure (point) in a chart. Hence it's a common structure around containers that can be plotted.

# Usability
Something really important IMHO is to keep as much as possible in the scala world, JSON and JS must be abstracted amap, unless you wanna do really fancy stuffs.

Of course, keeping reactivity is a must ^^.

**Comment by andypetrella** [comment]
@FRosner right, would be interesting indeed, I mean allowing non AMD javascript!

**Comment by FRosner** [comment]
Ok we will look into it in more detail this weekend. We are currently sketching a plan for a first prototype.
